### PR TITLE
fix `NamedStructField should be rewritten in OperatorToFunction` in subquery regression (change `ApplyFunctionRewrites` to use TreeNode API

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/10029

## Rationale for this change
1. Fix a regression

## What changes are included in this PR?
1. Properly recurse into subqueries that are embedded in expressions using the TreeNodeRewriter API
2. As a bonus avoids a bunch of copies

## Are these changes tested?

Yes with new tests
## Are there any user-facing changes?

Fix regression

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
@wiedld  and I found this while upgrading IOx. 

The code in this PR fixes the regression, but I think it could be substantially shorter (and do less copying) by using the TreeNode APIs.

I will work on that later today hopefully

**Comment by alamb** [comment]
I reworked this PR to use the TreeNode API in [a546e69](https://github.com/apache/arrow-datafusion/pull/10032/commits/a546e69124809547c9a27a0e38e020a48178ac15) and I am quite happy with how it looks now

File: datafusion/optimizer/src/analyzer/function_rewrite.rs [hunk 1]
```diff
@@ -37,86 +39,53 @@ impl ApplyFunctionRewrites {
...
         }
 
-        let mut expr_rewrite = OperatorToFunctionRewriter {
-            function_rewrites: &self.function_rewrites,
-            options,
-            schema: &schema,
-        };
+        let name_preserver = NamePreserver::new(&plan);
+
+        plan.map_expressions(|expr| {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/function_rewrite.rs:63]: using the new map_expressions API it is quite straightforward to rewrite these expressions (and it doesn't copy them!)

File: datafusion/optimizer/src/analyzer/function_rewrite.rs [hunk 2]
```diff
@@ -37,86 +39,53 @@ impl ApplyFunctionRewrites {
...
+                let mut result = Transformed::no(expr);
+                for rewriter in self.function_rewrites.iter() {
+                    result = result.transform_data(|expr| {
+                        rewriter.rewrite(expr, &schema, options)
+                    })?;
+                }
+                Ok(result)
+            })?;
 
-        plan.with_new_exprs(new_expr, new_inputs)
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/function_rewrite.rs:93]: this copies the plan + expressions

**Review by crepererum — APPROVED** [review]

**Comment by alamb** [comment]
Thanks for the review @crepererum  and for the help tracking this down @wiedld

**Pull request merged**